File: services/content-service/src/main/java/dev/animedia/contentservice/domain/shared/model/Page.java
package dev.animedia.contentservice.domain.shared.model;

import java.util.List;
import java.util.function.Function;

public record Page<T>(
    List<T> content,
    long totalElements,
    long totalPages,
    long pageNumber,
    long pageSize,
    boolean isFirst,
    boolean isLast,
    boolean hasNext,
    boolean hasPrevious
) {
    public <K> Page<K> changeContent(Function<T, K> mapper) {
        List<K> newContent = content.stream().map(mapper).toList();
        return changeContent(newContent);
    }

    public <K> Page<K> changeContent(List<K> newContent) {
        return new Page<>(
            newContent,
            totalElements,
            totalPages,
            pageNumber,
            pageSize,
            isFirst,
            isLast,
            hasNext,
            hasPrevious
        );
    }
}

File: services/social-service/src/core/Domain/Pagination/Entity/Page.php
<?php

namespace Core\Domain\Pagination\Entity;

readonly class Page
{
    public function __construct(
        public mixed $content,
        public int $pageNumber,
        public int $pageSize,
        public int $totalPages,
        public int $totalCount,
        public bool $hasPrev,
        public bool $hasNext
    ) {}

    public function changeContent(mixed $newContent, Page $page): Page
    {
        return new Page(
            content: $newContent,
            pageNumber: $page->pageNumber,
            pageSize: $page->pageSize,
            totalPages: $page->totalPages,
            totalCount: $page->totalCount,
            hasPrev: $page->hasPrev,
            hasNext: $page->hasNext
        );
    }
}
